temporian/core/operators/lag.py

```diff
@@ -84,21 +86,67 @@ def build_op_definition(cls) -> pb.OperatorDef:
 operator_lib.register_operator(LagOperator)
 
 
-def lag(event: Event, duration: Duration) -> Event:
-    if duration <= 0:
-        raise ValueError("duration must greater than zero")
+def _base_lag(
+    event: Event,
+    duration: Union[Duration, List[Duration]],
+    should_leak: bool = False,
+) -> Event:
+    """Base Lag function."""
+    if isinstance(duration, list):
+        used_duration = duration
+        if should_leak:
+            used_duration = [-d for d in duration]
+
+        if not all(isinstance(d, (Duration, int)) and d > 0 for d in duration):
+            raise ValueError(
+                "duration must be a list of positive Durations. Got"
+                f" {duration}, type {type(duration)}"
+            )
+
+        return [
+            LagOperator(event=event, duration=d).outputs()["event"]
+            for d in used_duration
+        ]
+
+    if not isinstance(duration, (Duration, int)) or duration <= 0:
+        raise ValueError(
+            f"duration must be a positive Duration. Got {duration}, type"
+            f" {type(duration)}"
+        )
+
+    used_duration = duration if not should_leak else -duration
 
     return LagOperator(
         event=event,
-        duration=duration,
+        duration=used_duration,
     ).outputs()["event"]
```

Comment:
This logic could be significantly simplified if we converted a single duration to a list first thing
```
if isinstance(duration, Duration):
    duration = [duration]

if not isinstance(duration, list):
    raise ...

<logic assuming duration is now a list of durations>
```